core/include/moveit/task_constructor/cost_terms.h

```diff
@@ -39,36 +39,80 @@
 #pragma once
 
 #include <moveit/task_constructor/storage.h>
+#include <moveit/task_constructor/utils.h>
 
 namespace moveit {
 namespace task_constructor {
-/// These structures all implement the Stage::CostTerm API and can be configured via Stage::setCostTerm()
+
+class CostTerm
+{
+public:
+	CostTerm(std::function<double(const SubTrajectory&)>);
+	CostTerm(std::function<double(const SubTrajectory&, std::string&)>);
+	CostTerm(std::nullptr_t);
+	CostTerm();
+
+	// is this a valid CostTerm?
+	operator bool() const;
+
+	double operator()(const SolutionBase& s, std::string& comment) const;
+
+	/** describes the supported types of solutions that should be forwarded to this CostTerm
+	 *
+	 * CostTerms that support more than `SubTrajectory`s should inherit and overwrite the internal supports_ flag
+	 */
+	enum class SolutionType
+	{
+		NONE = 0,
+		TRAJECTORY = 1 << 0,
+		SEQUENCE = 1 << 1,
+		WRAPPER = 1 << 2,
+		ALL = TRAJECTORY | SEQUENCE | WRAPPER
+	};
```

Comment:
I'm not convinced of this approach yet, but I didn't yet manage to review the commit `provide CostTerm class for all CostTerms`.